huaweicloud/data_source_huaweicloud_networking_secgroup.go

```diff
@@ -53,51 +57,129 @@ func DataSourceNetworkingSecGroup() *schema.Resource {
+	if err != nil {
+		return nil, fmt.Errorf("Error retriving security group list: %s", err)
+	}
+	if len(allGroups) == 0 {
+		return nil, fmt.Errorf("No sucurity group found, please change your search criteria and try again.")
+	}
+	logp.Printf("[DEBUG] The retrieved group list is: %v", allGroups)
+
+	filter := map[string]interface{}{
+		"ID":   d.Get("secgroup_id"),
+		"Name": d.Get("name"),
+	}
+	filterGroups, err := utils.FilterSliceWithField(allGroups, filter)
+	if err != nil {
+		return nil, fmtp.Errorf("Erroring filting security groups: %s", err)
+	}
+	if len(filterGroups) < 1 {
+		return nil, fmtp.Errorf("No Security Group found.")
+	}
+	if len(filterGroups) > 1 {
+		return nil, fmtp.Errorf("More than one Security Groups found.")
+	}
+
+	result := filterGroups[0].(v1Groups.SecurityGroup)
+	return &result, nil
+}
+
+func setV1SecurityGroupDetail(d *schema.ResourceData, config *config.Config, region string) diag.Diagnostics {
+	resp, err := getSecurityGroupByV1Client(d, config)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(resp.ID)
+
+	rules, err := flattenV1SecurityGroupRules(resp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	logp.Printf("[DEBUG] The retrieved rule list is: %v", rules)
+
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("name", resp.Name),
+		d.Set("description", resp.Description),
+		d.Set("enterprise_project_id", resp.EnterpriseProjectId),
+		d.Set("rules", rules),
+	)
+	return diag.FromErr(mErr.ErrorOrNil())
+}
+
+func getRuleListByGroupId(client *golangsdk.ServiceClient, groupId string) ([]map[string]interface{}, error) {
+	listOpts := v3Rules.ListOpts{
+		SecurityGroupId: groupId,
+	}
+	resp, err := v3Rules.List(client, listOpts)
+	if err != nil {
+		return nil, err
+	}
+	return flattenV3SecurityGroupRules(resp)
+}
+
 func dataSourceNetworkingSecGroupRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	config := meta.(*config.Config)
-	networkingClient, err := config.NetworkingV3Client(GetRegion(d, config))
+	region := GetRegion(d, config)
+	v3Client, err := config.NetworkingV3Client(region)
 	if err != nil {
-		return fmtp.DiagErrorf("Error creating HuaweiCloud networking client: %s", err)
+		return fmtp.DiagErrorf("Error creating HuaweiCloud networking v3 client: %s", err)
 	}
 
-	listOpts := groups.ListOpts{
-		ID:   d.Get("secgroup_id").(string),
-		Name: d.Get("name").(string),
-	}
-	if epsId := common.GetEnterpriseProjectID(d, config); epsId != "" {
-		listOpts.EnterpriseProjectId = epsId
-	} else {
-		listOpts.EnterpriseProjectId = "all_granted_eps"
+	listOpts := v3Groups.ListOpts{
+		ID:                  d.Get("secgroup_id").(string),
+		Name:                d.Get("name").(string),
+		EnterpriseProjectId: config.DataGetEnterpriseProjectID(d),
 	}
 
-	allSecGroups, err := groups.List(networkingClient, listOpts)
+	allSecGroups, err := v3Groups.List(v3Client, listOpts)
 	if err != nil {
+		if _, ok := err.(golangsdk.ErrDefault404); ok {
```

Comment:
is there a more accurate way to determine whether the V3 API is available？
error code 404 can be found in many situations

Reply:
In either case, the 404 will be verified using the API of the V1 version.